examples/misc/navigate_search_result_pages_caching_tokens.rb

```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env ruby
+# Encoding: utf-8
+#
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# GoogleAdsService.Search results are paginated but they can only be retrieved
+# in sequence starting by the first page. More details at
+# https://developers.google.com/google-ads/api/docs/reporting/paging.
+#
+# This example searches campaigns illustrating how GoogleAdsService.Search
+# result page tokens can be cached and reused to retrieve previous pages. This
+# is useful when you need to request pages that were already requested in the
+# past without starting over from the first page. For example, it can be used
+# to implement an interactive application that displays a page of results at a
+# time without caching all the results first.
+#
+# To add campaigns, run add_campaigns.rb.
+
+require 'optparse'
+require 'google/ads/google_ads'
+
+def navigate_search_result_pages_caching_tokens(customer_id)
+  # GoogleAdsClient will read a config file from
+  # ENV['HOME']/google_ads_config.rb when called without parameters
+  client = Google::Ads::GoogleAds::GoogleAdsClient.new
+
+  # The cache of page tokens. The key is the 0-indexed page number.
+  # The first page's token is always the empty string.
+  page_tokens = {
+    0 => ''
+  }
+
+  puts "Fetching page 1 to get metadata"
+  query = "SELECT campaign.id, campaign.name FROM campaign ORDER BY campaign.name LIMIT #{RESULTS_LIMIT}"
+  response = client.service.google_ads.search(
+    customer_id: customer_id,
+    query: query,
+    page_size: PAGE_SIZE,
+    return_total_results_count: true,
+  )
+
+  cache_next_page_token(page_tokens, response.page, 1)
```

Comment:
Should the argument here be `0`, since you're adding `1` in `cache_next_page_token()`?

Reply:
No, 1 is correct because this is for page 2 and cache_next_page_token is 0-indexed. More details on other comment.